### PR TITLE
Use right-valued reference passing for `TrackedMppDataPacketPtr`

### LINKED ISSUE #6302: TiFlash crash when using memory tracker
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
[2022/11/14 01:15:15.192 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 3163) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:406] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:414] ["Access: read."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.193 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Unknown si_code."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.201 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x5544c01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+89410561]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fb5c6568630\t<unknown symbol> [libpthread.so.0+63024]\n       0x161a5c6\tMemoryTracker::free(long) [tiflash+23176646]\n                \tdbms/src/Common/MemoryTracker.cpp:176\n       0x15cb8b2\tstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+22853810]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315\n       0x61c614e\tstd::__1::__vector_base<std::__1::shared_ptr<DB::TrackedMppDataPacket>, std::__1::allocator<std::__1::shared_ptr<DB::TrackedMppDataPacket> > >::~__vector_base() [tiflash+102523214]\n                \t/usr/local/bin/../include/c++/v1/vector:487\n       0x15dd622\tDB::HashPartitionWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::partitionAndEncodeThenWriteBlocks() [tiflash+22926882]\n                \tdbms/src/Flash/Mpp/HashPartitionWriter.cpp:122\n       0x618c222\tDB::ExchangeSenderBlockInputStream::readImpl() [tiflash+102285858]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.cpp:43\n       0x556f175\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+89583989]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75\n       0x556ee65\tDB::IProfilingBlockInputStream::read() [tiflash+89583205]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43\n       0x61929de\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+102312414]\n                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270\n       0x61924f6\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+102311158]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x16e8dfb\tDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+24022523]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x16ec343\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+24036163]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x16ec016\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+24035350]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x16ed5c2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+24040898]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fb5c6560ea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 4295) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:406] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:414] ["Access: read."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.203 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Unknown si_code."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.205 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x5544c01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+89410561]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fb5c6568630\t<unknown symbol> [libpthread.so.0+63024]\n       0x161a5c6\tMemoryTracker::free(long) [tiflash+23176646]\n                \tdbms/src/Common/MemoryTracker.cpp:176\n       0x15cb8b2\tstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+22853810]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315\n       0x61c614e\tstd::__1::__vector_base<std::__1::shared_ptr<DB::TrackedMppDataPacket>, std::__1::allocator<std::__1::shared_ptr<DB::TrackedMppDataPacket> > >::~__vector_base() [tiflash+102523214]\n                \t/usr/local/bin/../include/c++/v1/vector:487\n       0x15dd622\tDB::HashPartitionWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::partitionAndEncodeThenWriteBlocks() [tiflash+22926882]\n                \tdbms/src/Flash/Mpp/HashPartitionWriter.cpp:122\n       0x618c222\tDB::ExchangeSenderBlockInputStream::readImpl() [tiflash+102285858]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.cpp:43\n       0x556f175\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+89583989]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75\n       0x556ee65\tDB::IProfilingBlockInputStream::read() [tiflash+89583205]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43\n       0x61929de\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+102312414]\n                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270\n       0x61924f6\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+102311158]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x16e8dfb\tDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+24022523]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x16ec343\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+24036163]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x16ec016\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+24035350]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x16ed5c2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+24040898]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fb5c6560ea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.206 +08:00] [ERROR] [BaseDaemon.cpp:377] [########################################] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.206 +08:00] [ERROR] [BaseDaemon.cpp:378] ["(from thread 5074) Received signal Segmentation fault(11)."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.206 +08:00] [ERROR] [BaseDaemon.cpp:406] ["Address: NULL pointer."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.207 +08:00] [ERROR] [BaseDaemon.cpp:414] ["Access: read."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.207 +08:00] [ERROR] [BaseDaemon.cpp:426] ["Unknown si_code."] [source=BaseDaemon] [thread_id=5466]
[2022/11/14 01:15:15.207 +08:00] [ERROR] [BaseDaemon.cpp:570] ["\n       0x5544c01\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+89410561]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:221\n  0x7fb5c6568630\t<unknown symbol> [libpthread.so.0+63024]\n       0x161a5c6\tMemoryTracker::free(long) [tiflash+23176646]\n                \tdbms/src/Common/MemoryTracker.cpp:176\n       0x15cb8b2\tstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+22853810]\n                \t/usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315\n       0x61c614e\tstd::__1::__vector_base<std::__1::shared_ptr<DB::TrackedMppDataPacket>, std::__1::allocator<std::__1::shared_ptr<DB::TrackedMppDataPacket> > >::~__vector_base() [tiflash+102523214]\n                \t/usr/local/bin/../include/c++/v1/vector:487\n       0x15dd622\tDB::HashPartitionWriter<std::__1::shared_ptr<DB::MPPTunnelSet> >::partitionAndEncodeThenWriteBlocks() [tiflash+22926882]\n                \tdbms/src/Flash/Mpp/HashPartitionWriter.cpp:122\n       0x618c222\tDB::ExchangeSenderBlockInputStream::readImpl() [tiflash+102285858]\n                \tdbms/src/DataStreams/ExchangeSenderBlockInputStream.cpp:43\n       0x556f175\tDB::IProfilingBlockInputStream::read(DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*&, bool) [tiflash+89583989]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:75\n       0x556ee65\tDB::IProfilingBlockInputStream::read() [tiflash+89583205]\n                \tdbms/src/DataStreams/IProfilingBlockInputStream.cpp:43\n       0x61929de\tDB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::work(unsigned long, DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::WorkingInputs&) [tiflash+102312414]\n                \tdbms/src/DataStreams/ParallelInputsProcessor.h:270\n       0x61924f6\tstd::__1::__function::__func<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'(), std::__1::allocator<DB::ParallelInputsProcessor<DB::UnionBlockInputStream<(DB::StreamUnionMode)0, true>::Handler, (DB::StreamUnionMode)0>::process()::'lambda'()>, void ()>::operator()() [tiflash+102311158]\n                \t/usr/local/bin/../include/c++/v1/__functional/function.h:345\n       0x16e8dfb\tDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+24022523]\n                \tdbms/src/Common/ExecutableTask.h:52\n       0x16ec343\tDB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+24036163]\n                \tdbms/src/Common/DynamicThreadPool.cpp:101\n       0x16ec016\tDB::DynamicThreadPool::dynamicWork(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >) [tiflash+24035350]\n                \tdbms/src/Common/DynamicThreadPool.cpp:142\n       0x16ed5c2\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > >(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >), DB::DynamicThreadPool*&&, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&&)::'lambda'(auto&&...), DB::DynamicThreadPool*, std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> > > >(void*) [tiflash+24040898]\n                \t/usr/local/bin/../include/c++/v1/thread:291\n  0x7fb5c6560ea5\tstart_thread [libpthread.so.0+32421]"] [source=BaseDaemon] [thread_id=5466]
```
### 4. What is your TiFlash version? (Required)
master 6b393836ea60e5695692e59644e2f9f6aaa81c07

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #6302

Problem Summary:

### What is changed and how it works?
-  Use right-valued reference passing for `TrackedMppDataPacketPtr`
- Copy `TrackedMppDataPacketPtr` for broadcast because `TrackedMppDataPacket::switchMemTracker` will be called by multiple threads

The reason for the crash is
- `HashPartitionWriter` pass `TrackedMppDataPacketPtr`  to local tunnel  by left-valued reference. And now the `TrackedMppDataPacketPtr` hold the memory tracker of mpp task a.
- `LocalTunnel::readForLocal` call `TrackedMppDataPacket::switchMemoryTracker` and now the `TrackedMppDataPacketPtr` hold the memory tracker of mpp task b.
- And then the memory tacker throw exception like `RSS(Resident Set Size) much larger than limit`.
- mpp task b finished, and mpp task b is destroyed, as well as mpp task b's memory tracker.
- And then `TrackedMppDataPacketPtr` destroy in `HashPartitionWriter::partitionAndEncode` of mpp task a, and call the illegal memory tracker from mpp task b
- And the tiflash node crash

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp [hunk 1]
```diff
@@ -27,6 +27,16 @@ namespace DB
 {
 namespace tests
 {
+namespace
+{
+TrackedMppDataPacketPtr newDataPacket(const String & data)
+{
+    auto data_packet_ptr = std::make_shared<TrackedMppDataPacket>();
+    data_packet_ptr->getPacket().set_data(data);
```
Comment (windtalker) [inline dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:35]: Set data is not enough, most of the information is stored in chunks. Why not just copy the whole packet?

Comment (SeaRise) [reply dbms/src/Flash/Mpp/tests/gtest_mpptunnel.cpp:35]: Because the original ut only set data here

**Review by windtalker — APPROVED** [review]
LGTM

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**